### ISSUE TITLE
Max Sharpe additional constraints

### ISSUE THREAD

**Issue by robertmartin8** [robertmartin8/PyPortfolioOpt#92]
Currently, `max_sharpe` does not support additional constraints:

```python
# Works for min volatility
ef = EfficientFrontier(mu, S)
ef.add_constraint(lambda w: w >= 0.01)
ef.min_volatility()

# FAILS
ef = EfficientFrontier(mu, S)
ef.add_constraint(lambda w: w >= 0.01)
ef.max_sharpe()
```

This is _not_ an inherent problem with max_sharpe optimisation. This [blog post](http://yetanothermathprogrammingconsultant.blogspot.com/2016/08/portfolio-optimization-maximize-sharpe.html) clearly describes how to convert constraints into the maximum sharpe format. An example of this in cvxpy is below, where I am constraining the joint allocation to the first two stocks to be greater than or equal to 5% of the portfolio.

```python
y = cp.Variable(n)
k = cp.Variable()
risk_free_rate=0
variance = cp.quad_form(y, S.values)
constraints = [(mu.values - risk_free_rate).T * y == 1, 
               cp.sum(y) == k,
               k >= 0,
               y >= 0,
               y[0] + y[1] >= k* 0.05
              ]
opt = cp.Problem(cp.Minimize(variance),
                constraints)
opt.solve()
```

The problem lies in my choice of API for `add_objective`: it currently requires a lambda function.

```python
ef.add_objective(lambda w: w[0] + w[1] >= 0.05)
```

The problem I am facing is that the transformed `max_sharpe` constraint requires a `k` to be inserted into the constraint, e.g `lambda w: w[0] + w[1] >= k * 0.05`, and `k` has to be a cvxpy variable (the same object used in the optimisation problem).

Long story short, my implementation of `max_sharpe` is pretty terrible _because_ I'm trying to maintain a nice modular API. If you have any suggestions (@schneiderfelipe) I'd love to hear them. The problem essentially boils down to this: how do I convert a user-specified lambda function like `lambda w: w[0] + w[1] >= 0.05` into a constraint `w[0] + w[1] >= k * 0.05` on the backend? 

My sneaky trick, which I was momentarily very pleased with, was to pass `w / k` to the lambda function, i.e `w[0]/k + w[1]/k >= 0.05`, which should be mathematically equivalent to `w[0] + w[1] >= k * 0.05`. But cvxpy doesn't like this 😞

**Comment by robertmartin8** [comment]
I have potentially reduced this to a simpler problem: where it is possible to multiply a cvxpy constraint inequality by a constant. 

In which case I can take the nonlinear constraint: `w[0] + w[1] / k  >= 0.05` and multiply both sides by `k` to give `w[0] + w[1] >= 0.05 * k`, which certainly is linear. Only trouble is that cvxpy treats constraints differently to expressions, so I can't just multiply both sides. 

Might have to bite the bullet and ask StackOverflow

**Comment by robertmartin8** [comment]
Just [posted on SO](https://stackoverflow.com/questions/61338298/cvxpy-converting-a-nonlinear-constraint-to-an-equivalent-linear-constraint). I'm ready to get marked as a duplicate 😭

**Comment by schneiderfelipe** [comment]
That's a tough problem. One possibility would be to separate `lambda` and value, just like `scipy.optimize` optimizers do (e.g., `{"fun": lambda w: w[0] + w[1], "type": "ineq", "value": 0.05}`). This makes things more complicated, but allows you to manipulate stuff:

```python3
In [1]: a = lambda w: w[0] + w[1]                                               

In [2]: a([0, 1])                                                               
Out[2]: 1

In [3]: a([1, 1])                                                               
Out[3]: 2

In [4]: k = 10.0                                                                

In [5]: b = lambda wp, k: a(wp) * k                                             

In [6]: b([0, 1], k)                                                            
Out[6]: 10.0

In [7]: b([1, 1], k)                                                            
Out[7]: 20.0
```

By receiving things separately, you can manipulate either the "function side" or the "value side":

```python3
In [10]: c = lambda wp, k: a(wp) > k                                            

In [11]: c([5, 5], k)                                                           
Out[11]: False

In [12]: c([5, 6], k)                                                           
Out[12]: True
```

To be honest, I don't like what I just suggested (your API is clean and straightforward, my suggestions mess it up), but I can't see any other way around it.

**Comment by robertmartin8** [comment]
Good news – it turns out it is possible to decompose cvxpy constraints, so I can fix all of this on the backend. cvxpy constraints have an `args` parameter that gives you access to the numbers and variables involved, so I should be able to construct a new constraint with the appropriate `k` factor.
